include/ginkgo/core/base/lin_op.hpp

```diff
@@ -562,6 +562,112 @@ class Permutable {
 };
 
 
+/**
+ * The RowGatherable is an interface that allows to get the row gather result of
+ * a LinOp.
+ *
+ * @tparam IndexType  the type of the indices used to extract/gather the rows
+ */
+template <typename IndexType>
+class RowGatherable {
```

Comment:
Do we need a new Mixin here? row gather on CSR and Dense have two quite separate functions, and don't need to be made generic? Also the fact that std::unique_ptr is not covariant as a return type makes for a weird interface for the _linop overload.

Reply:
Or do you need this for some mixed-precision application?

Reply:
Yes, I think so? because the RowGatherer does not have value type information, I need this interface to call row_gather.
I thought CSR and Dense have the same interface for row_gather, doesn't it?
I do not understand the last sentence. could you describe it more detail?

Reply:
you can't override a std::unique_ptr<LinOp> function() with a std::unique_ptr<Dense> function(). Covariance means that if your class is more specialized, your return type can also be more specialized. This is true for pointers, but not for smart pointers.

Reply:
Good to know this! does the compiler show the override only partially message for this kind on pointers?
I think the return type is not considered in the function overload, so I still need to use two version for them?